Issue/wc 9372 local SKU product search

malinajirka opened the pull request:
This PR adds support for local SKU product search to the `getProducts` method. This will be useful for improving the Order Creation flow in the Woo app - https://github.com/woocommerce/woocommerce-android/issues/9372.

Props to @rossanafmenezes - we solved this issue together during a couple pair programming sessions.

malinajirka committed:
Pass skuOptionsParams to getProducts method

rossanafmenezes committed:
Add logic for SKU search in getProduct

malinajirka committed:
Merge branch 'trunk' into issue/wc-9372-local-sku-product-search

malinajirka committed:
Set default value for skuSearchOptions

malinajirka committed:
Fix detekt and checkstyle warnings

samiuelson commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt:
```diff
@@ -245,14 +247,28 @@ object ProductSqlUtils {
+                }
+                SkuSearchOptions.ExactSearch -> {
+                    queryBuilder.beginGroup()
+                        .equals(WCProductModelTable.SKU, searchQuery.lowercase())
```
Why do we need to lowercase the query here?

malinajirka replied:
Great question - I wanted to return to this but I forgot. 

Using `lowerCase` here doesn't give us any benefit at all, since we are converting only the searchQuery but not the data in the database - I removed it in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2789/commits/ef80d315afa38f3a553f2a4384cd4c0db0cb4f6f.

Unfortunately, it seems WellSQL doesn't support COLLATE NOCASE, so I can't think of an elegant solution. For SkuSearchOptions.ExactSearch, we'll be doing a case sensitive search - which I guess might be expected. This method with ExactSearch is not being used at the moment in the Woo app.

samiuelson commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt:
```diff
@@ -245,14 +247,28 @@ object ProductSqlUtils {
+                }
+                SkuSearchOptions.PartialMatch -> {
+                    queryBuilder.beginGroup()
+                        .contains(WCProductModelTable.SKU, searchQuery)
```
Is the `contains()` method case-sensitive?

malinajirka replied:
Nope, `contains` in SQLite is case insensitive, AFAIK. (eg https://database.guide/how-to-make-sqlites-like-operator-case-sensitive/)

samiuelson reviewed (commented):
Thanks for your work. Looks good overall! 

I left some questions about parts that were unclear to me.

malinajirka committed:
Remove confusing 'lowerCase' conversion

WellSQL doesn't support collate nocase. Using 'lowerCase' doesn't give us any benefit since it converts only the input query not the data in the database.

malinajirka committed:
Add comment

malinajirka commented:
Thanks so much for the review @samiuelson, it's ready for another round.

samiuelson reviewed (approved):
Looks good! :shipit:
